samples/nrf9160/http_application_fota/src/main.c

```diff
@@ -0,0 +1,220 @@
+			enum download_client_evt event, u32_t status);
+
+
+static struct download_client dfu = {
+	.host = CONFIG_HOST,
+	.resource = CONFIG_RESOURCE,
+	.callback = app_download_client_event_handler
+};
+
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	LOG_ERR("%s = %lu\n", __func__, err);
+
+	/* Read the fault register for details. */
+	volatile uint32_t fault_reason = *((volatile uint32_t *)0x4002A614);
+
+	__ASSERT(false, "%s, reason %lu, reset application\n",
+					__func__, fault_reason);
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	LOG_ERR("%s = %lu\n", __func__, err);
+
+	/* Read the fault register for details. */
+	volatile uint32_t fault_reason = *((volatile uint32_t *)0x4002A614);
+
+	__ASSERT(false, "%s, reason %lu, reset application\n",
+					 __func__, fault_reason);
+}
+
+
+/**@brief Initialize application. */
+static void app_dfu_init(void)
+{
+	flash_dev = device_get_binding(DT_FLASH_DEV_NAME);
+
+	__ASSERT(flash_dev != 0, "Nordic nRF flash driver was not found!\n");
+
+	int retval = download_client_init(&dfu);
+
+	__ASSERT(retval == 0, "download_client_init() failed, err %d", retval);
+}
+
+
+/**@brief Start transfer of the file. */
+static void app_dfu_transfer_start(void)
+{
+	int retval = download_client_connect(&dfu);
+
+	__ASSERT(retval == 0, "download_client_connect() failed, err %d",
+		retval);
+
+	retval = download_client_start(&dfu);
+	__ASSERT(retval != -1, "download_client_start() failed, err %d",
+		retval);
+}
+
+
+static int app_download_client_event_handler(
+	struct download_client *const dfu,
+	enum download_client_evt event, u32_t error)
+{
+	int err;
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		static u32_t address = CONFIG_FLASH_WRITE_OFFSET;
+
+		/* Erase page(s) here: */
+		flash_write_protection_set(flash_dev, false);
+		err = flash_write(flash_dev, address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			__ASSERT(false,
+					"Flash write error %d at address %08x\n",
+					err, address);
+		}
+		address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
+		LOG_INF("DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE");
+		break;
+
+	case DOWNLOAD_CLIENT_EVT_ERROR: {
+		LOG_ERR("DFU error");
+		download_client_disconnect(dfu);
+		__ASSERT(false, "Something went wrong, please restart "
+						"the application\n");
+		break;
+	}
+	default:
+		break;
+	}
+	return 0;
+}
+
+
+/**@brief Turn on LED0 if APP_VERSION is 1 and LED1 otherwise*/
+static void led_app_version(void)
+{
+	struct device *dev;
+
+	dev = device_get_binding(LED_PORT);
+	__ASSERT(dev != 0, "Nordic nRF GPIO driver was not found!\n");
+
+	gpio_pin_configure(dev, LED0_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED0_GPIO_PIN, 1);
+
+#if APP_VERSION == 2
```

Comment:
Is APP_VERSION supposed to be configurable? As it stands now a user will have to go into the example and change APP_VERSION in the source code when preparing an "updated" app which blinks another led than the initial DFU example.

I see two solution to this:
1. Create two build targets so that it outputs `zephyr.hex` and `zephyr_update.hex`
2. Make it configurable in the samples Kconfig where you can increment the version number with a range 1-2

Reply:
Yes, will do that.

Reply:
config APPLICATION_VERSION added to Kconfig now.